### PR TITLE
React Native support

### PR DESCRIPTION
**Pull request by dbrrt**
I've splitted the meteor-apollo-accounts lib into another folder, because we can't really import conditionally AsyncStorage

### CODE REVIEW

**Comment by nicolaslopezj** [comment]
I don't like having 2 separate packages..

We could use something like https://github.com/feathersjs/feathers-localstorage/

**Comment by dbrrt** [comment]
Do you know how to detect if an application is React-Native or not ?
I'd suggest using this package to make it "hybrid" : https://www.npmjs.com/package/react-native-storage, but if you want something working with both, we'll need to find a way to adapt the logic depending if it's react or react native.

**Comment by nicolaslopezj** [comment]
Check here https://github.com/nicolaslopezj/simple-react-form/blob/master/src/Form/index.js#L416

**Comment by dbrrt** [comment]
So it works great, thanks !
I'm still stuck with the conditional import of AsyncStorage. Do you have any clue for that?

**Comment by nicolaslopezj** [comment]
Mmm I haven't done that. But we could to use an external library that does that jobs

**Comment by dbrrt** [comment]
Do you mind if the react native client is Apollo only (for now)? We could eventually store the local data using Redux and something like that https://github.com/rt2zz/redux-persist ? 
But we'll face the exact same problem because AsyncStorage is needed and seems to be the only thing to store data onto react-native. 
And having two libraries for different type of client is ugly tho...

**Comment by janikvonrotz** [comment]
Have you tried something like:

```
import AsyncStorage from `react-native`

if(reactNative){
  LocalStorage = AsyncStorage
}
export LocalStorage
```

Or why not offer the possiblity to inject a local storage?

```
import AsyncStorage from `react-native`
import initApp from `apollo-accounts`

initApp(AsyncStorage)
```

**Comment by dbrrt** [comment]
@janikvonrotz 
Because the library has to be compatible with React and React-Native, but an exclusive ReactJS willn't want to install React-Native to have it working.
I like the injection of Storage, might do the job, I'll try it after work.

**Comment by nicolaslopezj** [comment]
I had not thought using a react native solution would make this package react only. As @dbrrt this package should not be exclusive for react. I like @janikvonrotz idea

**Comment by dbrrt** [comment]
@janikvonrotz   I tried the following code without any success, it crashes : ReferenceError: navigator is not defined and the require is executed automatically and makes the application crashing on a React Apollo instance without React Native installed. :

``` javascript
// Doesn't run
var LocalStorage;

var navigatorUser = navigator;

if(navigatorUser && navigatorUser.product === 'ReactNative') {

    var RNative = require("react-native");

    var {
        AsyncStorage
    } = RNative;

    LocalStorage = AsyncStorage;
}
else {
  LocalStorage = global.localStorage;
}
```
The injection remains the last possibility to make it hybrid, but it's not super clean.
It'd require having LocalStorage available globally and be initiated once by the client.
Any help welcomed ;)

**Comment by nicolaslopezj** [comment]
I think something like this could be the solution

```js
let tokenStore = {
  set: async function ({userId, token, tokenExpires}) {
    global.localStorage['Meteor.userId'] = userId
    global.localStorage['Meteor.loginToken'] = token
    global.localStorage['Meteor.loginTokenExpires'] = tokenExpires
  },
  get: async function ({userId, token, tokenExpires}) {
    return {
      userId: global.localStorage['Meteor.userId'], 
      token: global.localStorage['Meteor.loginToken'], 
      tokenExpires: global.localStorage['Meteor.loginTokenExpires']
    }
  }
}

export const setTokenStore = function (newStore) {
  tokenStore = newStore
}

export const storeLoginToken = function (userId, token, tokenExpires) {
  tokenStore.set({userId, token, tokenExpires})
  tokenDidChange()
}
```


And in every React Native app people should do:

```js
// Init Meteor Accounts
import {AsyncStorage} from 'react-native'
import {setTokenStore} from 'meteor-apollo-accounts'

setTokenStore({
  set: async function ({userId, token, tokenExpires}) {
    await AsyncStorage.setItem('Meteor.userId', userId)
    await AsyncStorage.setItem('Meteor.loginToken', token)
    await AsyncStorage.setItem('Meteor.loginTokenExpires', tokenExpires)
  },
  get: async function () {
    return {
      userId: await AsyncStorage.getItem('Meteor.userId'), 
      token: await AsyncStorage.getItem('Meteor.loginToken'), 
      tokenExpires: await AsyncStorage.getItem('Meteor.loginTokenExpires')
    }
  }
})
```

**Comment by dbrrt** [comment]
It works flawlessly! Bravo @nicolaslopezj ! I updated my repository for the PR.

File: meteor-apollo-accounts/src/createUser.js [hunk 1]
```diff
@@ -2,11 +2,11 @@ import hashPassword from './hashPassword'
 import gql from 'graphql-tag'
 import {storeLoginToken} from './store'
 
-export default async function ({username, email, password, profile}, apollo) {
```
Comment (nicolaslopezj) [inline meteor-apollo-accounts/src/createUser.js]: Why removing profile functionality?

File: meteor-apollo-accounts/src/index.js [hunk 2]
```diff
@@ -9,9 +9,9 @@ import resetPassword from './resetPassword'
 import verifyEmail from './verifyEmail'
 import loginWithFacebook from './oauth/loginWithFacebook'
 import loginWithGoogle from './oauth/loginWithGoogle'
-import loginWithLinkedIn from './oauth/loginWithLinkedIn'
```
Comment (nicolaslopezj) [inline meteor-apollo-accounts/src/index.js]: Here again removing linkeding

File: meteor-apollo-accounts/src/store.js [hunk 3]
```diff
@@ -1,16 +1,39 @@
...
 
-export const getLoginToken = function () {
-  return global.localStorage['Meteor.loginToken']
+let tokenStore = {
+  set: async function ({userId, token, tokenExpires}) {
+    global.localStorage['Meteor.userId'] = userId
+    global.localStorage['Meteor.loginToken'] = token
+    global.localStorage['Meteor.loginTokenExpires'] = tokenExpires.toString()
+  },
+  get: async function ({userId, token, tokenExpires}) {
```
Comment (nicolaslopezj) [inline meteor-apollo-accounts/src/store.js]: Why passing variables?

File: meteor-apollo-accounts/src/store.js [hunk 4]
```diff
@@ -1,16 +1,39 @@
...
+  tokenStore = newStore
+}
+
+export const storeLoginToken = function (userId, token, tokenExpires) {
+  tokenStore.set({userId, token, tokenExpires})
+  tokenDidChange()
+}
+
+export const getLoginToken = async function () {
+  return tokenStore.get({token}) || null
```
Comment (nicolaslopezj) [inline meteor-apollo-accounts/src/store.js]: This should be:

```js
tokenStore.get().token
```

File: meteor-apollo-accounts/src/createUser.js (hunk 1 above)
Comment (dbrrt) [inline meteor-apollo-accounts/src/createUser.js]: My branch was too old, I'll update that.

File: meteor-apollo-accounts/src/index.js (hunk 2 above)
Comment (dbrrt) [inline meteor-apollo-accounts/src/index.js]: Same issue than with profile removing

File: meteor-apollo-accounts/src/store.js (hunk 3 above)
Comment (dbrrt) [inline meteor-apollo-accounts/src/store.js]: I've been too fast on copy-paste, I removed that after.

**Comment by dbrrt** [comment]
@nicolaslopezj  I fixed my commit. Let me know if you see things to edit.

File: meteor-apollo-accounts/src/store.js [hunk 5]
```diff
@@ -1,16 +1,39 @@
...
+}
+
+export const getLoginToken = async function () {
+  return tokenStore.get().token || null
 }
 
-export const getUserId = function () {
-  return global.localStorage['Meteor.userId'] || null
+export const getUserId = async function () {
+  return tokenStore.get().userId || null
```
Comment (nicolaslopezj) [inline meteor-apollo-accounts/src/store.js]: return await tokenStore.get().userId || null

**Comment by nicolaslopezj** [comment]
Please update the changelog too

File: meteor-apollo-accounts/src/store.js (hunk 5 above)
Comment (dbrrt) [inline meteor-apollo-accounts/src/store.js]: Edited

**Comment by nicolaslopezj** [comment]
Looks great! did you test it in react native?

**Comment by dbrrt** [comment]
@nicolaslopezj, As I said yes,  I tried it with React Native (0.39.2) onto iPhone simulator (IOS 10.2) on Mac OS Sierra. It seems to work normally.

**Comment by nicolaslopezj** [comment]
Please add to the README instructions to use it on RN

**Comment by barrada3** [comment]
I'll add this after work, but except the storing part, it works the same on React and React Native.

**Pull request merged**